File: sorts/merge_sort.py
from random import randint
from numpy import array
from utils import check_correctness


def merge(left, right):
    res = []
    i = j = 0

    for k in range(len(left) + len(right)):
        if i == len(right):
            return res + left[j:]

        if j == len(left):
            return res + right[i:]

        if right[i] <= left[j]:
            res.append(right[i])
            i = i + 1
        else:
            res.append(left[j])
            j = j + 1

    return res


def merge_sort(arr):
    if len(arr) >= 2:
        middle = len(arr) // 2
        left = merge_sort(arr[:middle])
        right = merge_sort(arr[middle:])

        return merge(left, right)
    else:
        return arr


def main():
    arr = [randint(1, 100) for i in range(10)]
    arr = merge_sort(arr)

    if check_correctness(arr):
        print("MERGE SORT OK")
    else:
        print("MERGE SORT FAIL")
                
                
if __name__ == "__main__":
    main()

File: sorts/merge_sort.hs
mergeSort :: Ord a => [a] -> [a]
mergeSort (x:[]) = [x]
mergeSort xs = merge (mergeSort left) (mergeSort right) 
    where middle = length xs `div` 2
          left = take middle xs
          right = drop middle xs

merge :: Ord a => [a] -> [a] -> [a]
merge left []    = left
merge [] right   = right
merge left right = 
    if head left < head right 
        then head left : merge (tail left) right
        else head right : merge left (tail right)

main :: IO ()
main = do
    print $ mergeSort [5, 3, 6, 10, -1, 15, 14, -2, 5]
